packages/core-transaction-pool/src/processor.ts

```diff
@@ -1,436 +1,57 @@
-                .getActivatedHandlerForData(transaction);
-            return handler.canEnterTransactionPool(transaction, this.transactionPool, this);
-        } catch (error) {
-            if (error instanceof Errors.InvalidTransactionTypeError) {
-                this.pushError(
-                    transaction,
-                    "ERR_UNSUPPORTED",
-                    `Invalidating transaction of unsupported type '${Enums.TransactionType[transaction.type]}'`,
-                );
-            } else {
-                this.pushError(transaction, "ERR_UNKNOWN", error.message);
-            }
-        }
-
-        return false;
-    }
-
-    /**
-     * @private
-     * @returns {Promise<void>}
-     * @memberof Processor
-     */
-    private async addTransactionsToPool(): Promise<void> {
-        const { notAdded }: TransactionsProcessed = await this.transactionPool.addTransactions(
-            Array.from(this.accept.values()),
-        );
-
-        for (const item of notAdded) {
-            Utils.assert.defined<Interfaces.ITransaction>(item.transaction);
-
-            const transaction: Interfaces.ITransaction = item.transaction;
-
-            Utils.assert.defined<string>(transaction.id);
-
-            const id: string = transaction.id;
-
-            this.accept.delete(id);
+                await this.pool.add(transaction);
+                this.accept.push(transaction.id);
+                if (dynamicFee.broadcast) {
+                    broadcastable.push(transaction);
+                    this.broadcast.push(transaction.id);
+                }
+            } catch (error) {
+                if (!this.errors) {
+                    this.errors = [];
+                }
+                this.errors.push(error.message);
 
-            if (item.type !== "ERR_POOL_FULL") {
-                this.broadcast.delete(id);
+                if (error.type === "ERR_EXCEEDS_MAX_COUNT") {
+                    this.excess.push(transaction.id);
+                } else {
+                    this.invalid.push(transaction.id);
+                }
             }
-
-            Utils.assert.defined<string>(item.type);
-            Utils.assert.defined<string>(item.message);
-
-            this.pushError(transaction.data, item.type, item.message);
-        }
-    }
-
-    /**
-     * @private
-     * @memberof Processor
-     */
-    private printStats(): void {
-        const stats: string = ["accept", "broadcast", "excess", "invalid"]
-            .map(prop => `${prop}: ${this[prop] instanceof Array ? this[prop].length : this[prop].size}`)
-            .join(" ");
-
-        if (Object.keys(this.errors).length > 0) {
-            this.logger.debug(JSON.stringify(this.errors));
         }
 
-        this.logger.info(`Received ${Utils.pluralize("transaction", this.transactions.length, true)} (${stats}).`);
+        await this.networkMonitor.broadcastTransactions(broadcastable);
```

Comment:
This should be moved out so that the sole responsibility of the processor is to decide the state of a transaction (accept, broadcast, excess, invalid, error). The broadcasting is a separate task, the processor should only provide the data that is needed for it.

Reply:
I disagree. Why i can't add to pool and broadcast instead of requiring caller to do that?

Reply:
The reasoning is both responsibility and testability. The whole function right now is coupled to the network monitor from the P2P package just for broadcasting which means during tests you have to mock the implementation or fake a response.

99% of this function is attached to the pool, about deciding if transactions are valid and adding them to the pool and then there is this 1 unrelated action in the shape of the P2P call. The transaction pool itself should be unaware of the existence of the P2P layer, it only cares and knows about what transactions are and if they are valid. What the caller wants to do with the transactions after it has handled them is not of any concern for it.

Reply:
This is actually the reason I think processor should not be part of pool. But more general service on top of pool and p2p (or other broadcast service).

If instead instructions are returned back (previous processor), then someone still has to execute them. There are currently two places that do that, and they did that differently before #3514. P2P was incorrectly(?) refusing the whole batch if single transaction failed to apply, but API broadcasted whatever applied. 

It's _very_ easy to make mistake and accidently do the wrong thing. Processor that was implemented in #3514 isn't simple and probably has errors, but everything is contained within it.

Reply:
The rejection of everything via P2P even if just 1 transaction is invalid is on purpose as a precaution to prevent someone from sending junk directly to a node via P2P and it causing problems. P2P in general has some different behaviours than the API because the API only cares about taking any data and broadcasting while P2P double checks that everything is 100% valid.